File: 2021/10/30/main.go
package main

func longestDupSubstring(s string) string {
	res := ""
	isExist := func(n int) bool {
		seen := make(map[string]bool, len(s)-n+2)
		for i := 0; i+n <= len(s); i++ {
			sub := s[i : i+n]
			if seen[sub] {
				res = sub
				return true
			}
			seen[sub] = true
		}
		return false
	}

	l, r := 0, len(s)
	for l < r {
		m := r - (r+l)/2
		if isExist(m) {
			l = m
		} else {
			r = m - 1
		}
	}

	return res
}

File: 2021/10/30/main.py
class Solution:
    def longestDupSubstring(self, s: str) -> str:
        self.res = ""
        def isExist(n):
            seen = set()
            for i in range(len(s)-n+1):
                sub = s[i:i+n]
                if sub in seen:
                    self.res = sub
                    return True
                else:
                    seen.add(sub)
            return False

        l, r = 0, len(s)
        while l < r:
            m = r - (r - l) // 2
            if isExist(m):
                l = m
            else:
                r = m - 1
        return self.res